experimental/sql/dialects/rel_ssa.py

```diff
@@ -0,0 +1,271 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from typing import Any, Union, List
+from xdsl.ir import Block, Region, Operation, SSAValue, ParametrizedAttribute, Data, MLContext, Attribute
+from xdsl.dialects.builtin import StringAttr, ArrayAttr, ArrayOfConstraint, IntegerAttr
+from xdsl.irdl import AttributeDef, OperandDef, ResultDef, RegionDef, SingleBlockRegionDef, irdl_attr_definition, irdl_op_definition, ParameterDef, AnyAttr, VarOperandDef, builder
+
+# This file contains the relational ssa dialect, a dialect that expresses
+# relational queries using SSA. The dialect consists of two types of operations:
+# Expressions and Operators. Expressions work on single values, whereas
+# Operators work on Bags. The Column and Yield Expressions bridge the gap to
+# convert from one to the other.
+
+
+@irdl_attr_definition
+class DataType(ParametrizedAttribute):
+  """
+  Models a datatype in a relational ssa query.
+  """
+  name = "rel_ssa.datatype"
+
+
+@irdl_attr_definition
+class Int32(DataType):
+  """
+  Models a int32 type in a relational ssa query.
+
+  Example:
+
+  ```
+  !rel_ssa.int32
+  ```
+  """
+  name = "rel_ssa.int32"
+
+
+@irdl_attr_definition
+class String(DataType):
+  """
+  Models a string type in a relational ssa query, that can be either nullable or
+  not.
+
+  Example:
+
+  ```
+  !rel_ssa.string<0: !i1>
+  ```
+  """
+  name = "rel_ssa.string"
+
+  nullable = ParameterDef(IntegerAttr)
+
+  @staticmethod
+  @builder
+  def get(val: Union[int, IntegerAttr]) -> 'String':
+    if isinstance(val, IntegerAttr):
+      return String([val])
+    return String([IntegerAttr.from_int_and_width(val, 1)])
+
+
+@irdl_attr_definition
+class Boolean(ParametrizedAttribute):
+  """
+  Models a type that can either be true or false to, e.g., show whether a tuple
+  is part of the result or not.
+
+  Example:
+
+  '''
+  !rel_ssa.bool
+  '''
+  """
+  name = "rel_ssa.bool"
+
+
+@irdl_attr_definition
+class Bag(ParametrizedAttribute):
+  """
+  Models a bag in a relational ssa query. The exact schema of the bag is part of
+  the type itself.
+
+  Example:
+
+  '''
+  !rel_ssa.bag<[!rel_ssa.int32]>
+  '''
+  """
+  name = "rel_ssa.bag"
+
+  schema = ParameterDef(ArrayOfConstraint(DataType))
+
+  @staticmethod
+  @builder
+  def get(types: list[DataType]) -> 'Bag':
+    return Bag([ArrayAttr.from_list(types)])  #type: ignore
+
+
+class Expression(Operation):
+  """
+  Interface class for all Expressions, i.e., operations that work on single
+  values.
+  """
+  ...
+
+
+@irdl_op_definition
+class Column(Expression):
+  """
+  References a specific column with name `col_name` and returns its value.
```

Comment:
I think it would be helpful to briefly talk about the scope of these names. Here is probably a good place. In this dialect, names refer to the schema of the encompassing operator -- and *never* an operator that precedes it like in the Ibis dialect.